File: logstash-core/lib/logstash/util/apm.rb
module LogStash; module Util; module Apm
  extend self

  require 'java'

  java_import 'io.opentelemetry.api.GlobalOpenTelemetry'
  java_import 'io.opentelemetry.api.trace.Span'
  java_import 'io.opentelemetry.api.trace.Tracer'
  java_import 'io.opentelemetry.context.Context'
  java_import 'io.opentelemetry.context.propagation.TextMapPropagator'
  java_import 'io.opentelemetry.context.propagation.TextMapSetter'

  def with_span(name)
    begin
        puts "Measuring #{name}"
        parent_span = Java::io.opentelemetry.api.trace.Span.current
        puts "the parent span is #{parent_span}"
        span = Java::co.elastic.logstash.api.APM.startTrace(name)
        scope = span.make_current
#       parent_span = Java::co.elastic.apm.api.ElasticApm.currentSpan
#       span = parent_span.startSpan
#       span.setName(name)
#       scope = span.activate
      yield
    ensure
        puts "the scope is #{scope}, the span is #{span}"
        scope.close unless scope.nil?
        span.end unless span.nil?
        puts "Measured #{name}"
#       scope.close
#       span.end
    end
  end

  def add_traceparent(hash)
      propagator = Java::co.elastic.logstash.api.APM.getPropagator
      propagator.inject(Java::io.opentelemetry.context.Context.current(), hash, lambda { |carrier, key, value| carrier.put(key, value) })
      puts "hash is #{hash}"
      hash
  end

end;end;end

File: logstash-core/src/main/java/co/elastic/logstash/api/APM.java
package co.elastic.logstash.api;

import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.api.metrics.LongCounter;
import io.opentelemetry.api.metrics.Meter;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanBuilder;
import io.opentelemetry.api.trace.SpanKind;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.context.Context;
import io.opentelemetry.context.Scope;
import io.opentelemetry.context.propagation.TextMapGetter;
import io.opentelemetry.context.propagation.TextMapPropagator;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

public class APM {

    private static OpenTelemetry openTelemetry;
    private static Tracer tracer;
    private static Meter meter;
    private final Map<String, Context> spans = new ConcurrentHashMap<>();
    private static LongCounter counter;
    private static TextMapPropagator propagator;

    public static void setup(){
        openTelemetry = GlobalOpenTelemetry.get();
        tracer = openTelemetry.tracerBuilder("logstash")
                .setInstrumentationVersion("unknown")
                .setSchemaUrl("https://opentelemetry.io/schemas/1.21.0")
                .build();
        meter = GlobalOpenTelemetry.getMeter("my_meter");
        counter = meter.counterBuilder("my_counter").build();
        propagator = GlobalOpenTelemetry.getPropagators().getTextMapPropagator();
//        tracer = openTelemetry.getTracer("logstash", "8.14.0");
    }

    public static OpenTelemetry getOTel(){
        return openTelemetry;
    }

    public static Tracer getTracer(){
        return tracer;
    }

    public static TextMapPropagator getPropagator(){
        return propagator;
    }

    public static Span startTrace(String name){
        var spanBuilder = tracer.spanBuilder(name);
        spanBuilder.setSpanKind(SpanKind.INTERNAL);
        return spanBuilder.startSpan();
    }

    public static void stopTrace(Span span){
        span.end();
    }

}
